### PR TITLE
UI Fade in and out code and implementation

### PR DESCRIPTION
**Pull request by ZombieNub**
This implements the UI fader to black card using events and scriptable objects. To use, add a FadeChannelSO to the relevant script, assign the scriptable object, the call either Fade, FadeIn, or FadeOut. Has parameters for duration and target color. This is usable anywhere and at any time, but contains no safeguards besides preventing multiple fades from happening at the same time.

Note: This modifies the PersistentManagers scene and adds an Image UI element that it depends on.

Forum Thread: [https://forum.unity.com/threads/ui-fader-to-black.1038013/](https://forum.unity.com/threads/ui-fader-to-black.1038013/)

I would like to thank forum user [Swerik ](https://forum.unity.com/members/swerik.5425601/)for proposing the implementation plan I ended up using.

### CODE REVIEW

File: UOP1_Project/Assets/Scripts/Events/ScriptableObjects/UI/FadeChannelSO.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	/// <summary>
+	/// Generic fade function. Communicates with <seealso cref="FadeManager.cs"/>.
+	/// </summary>
+	/// <param name="fadeIn">If true, the rectangle fades in. If false, the rectangle fades out.</param>
+	/// <param name="duration">How long it takes to the image to fade in/out.</param>
+	public void Fade(bool fadeIn, float duration)
+	{
+		if (OnEventRaised != null)
+			OnEventRaised.Invoke(fadeIn, duration, new Color(0, 0, 0, 1));
+	}
```
Comment (TheSwerik) [inline UOP1_Project/Assets/Scripts/Events/ScriptableObjects/UI/FadeChannelSO.cs:31]: You can give parameters a default value so that you don't have to write the same method multiple times.
Here an example:
```suggestion
	public void Fade(bool fadeIn, float duration, Color color = default(Color))
	{
		if (color == default(Color)) color = Color.black;
		if (OnEventRaised != null)
			OnEventRaised.Invoke(fadeIn, duration, color);
	}
```
This would work for the FadeIn-Methods as well

FYI: default(Color) == new Color(0, 0, 0, 0) and that's the only problem: If you want to fade in to or out from transparent but I don't see how that makes any sense so I would ignore that

File: UOP1_Project/Assets/Scripts/UI/FadeManager.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		while (totalTime <= duration)
+		{
+			totalTime += Time.deltaTime;
+			_imageComponent.color = Color.Lerp(oldColor, new Color(0, 0, 0, 0), totalTime / duration); // Sets the image's color to a mixture between the old color and  total transparency, and interpolates based on the amount of time to completion.
+			yield return null;
+		}
+		_imageComponent.color = new Color(0, 0, 0, 0); // Here to guarentee the image is fully transparent at the end of the loop.
+		IsCurrentlyFading = false;
+		yield return null;
+	}
```
Comment (TheSwerik) [inline UOP1_Project/Assets/Scripts/UI/FadeManager.cs:56]: Here I said [in the Thread](https://forum.unity.com/threads/ui-fader-to-black.1038013/#post-6728329) that I would like to keep the ability to choose a FadeOut starting color

you could do this with a default value so that if you don't put anything as a parameter it automatically chooses the old color:
```suggestion
	private IEnumerator FadeOutEnum(float duration, Color color = default)
	{
		if (color == default) color = _imageComponent.color; // Stores the old color of the image component, as we can't assume the image will always be black, if no color is specified.
		float totalTime = 0f; // Total amount of time this coroutine has taken. Determines when the fadeout will end and what color the imageComponent should be at every frame.
		while (totalTime <= duration)
		{
			totalTime += Time.deltaTime;
			_imageComponent.color = Color.Lerp(color, new Color(0, 0, 0, 0), totalTime / duration); // Sets the image's color to a mixture between the old color and  total transparency, and interpolates based on the amount of time to completion.
			yield return null;
		}
		_imageComponent.color = new Color(0, 0, 0, 0); // Here to guarentee the image is fully transparent at the end of the loop.
		IsCurrentlyFading = false;
		yield return null;
	}
```

Comment (TheSwerik) [reply UOP1_Project/Assets/Scripts/UI/FadeManager.cs:56]: You would need to change it in the FadeChannelSO.cs as well so that You can choose a FadeOut starting color there as well.

Comment (TheSwerik) [reply UOP1_Project/Assets/Scripts/UI/FadeManager.cs:56]: and you'd need to change it in this file line 96 `StartCoroutine(FadeOutEnum(duration));` aswell

Comment (ZombieNub) [reply UOP1_Project/Assets/Scripts/UI/FadeManager.cs:56]: Your two comments make sense. I'll implement the changes and update the PR.

File: UOP1_Project/Assets/Scripts/UI/FadeManager.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+public class FadeManager : MonoBehaviour
+{
+	[Header("Fadeout Event")]
+	[SerializeField] private FadeChannelSO _fadeChannelSO = default;
+
+	[Header("Fadeout Canvas")]
+	public GameObject _fadeoutCanvas = default;
+
+	private Image _imageComponent;
```
Comment (TheSwerik) [inline UOP1_Project/Assets/Scripts/UI/FadeManager.cs:14]: why grab the canvas when you could just grab the image itself?
```suggestion
	[SerializeField] private Image imageComponent;
```
Just drag the FadeOut Rectangle in to get the image
then you can also remove the Start() method

File: UOP1_Project/Assets/Scripts/UI/FadeManager.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+
+public class FadeManager : MonoBehaviour
+{
+	[Header("Fadeout Event")]
+	[SerializeField] private FadeChannelSO _fadeChannelSO = default;
+
+	[Header("Fadeout Canvas")]
```
Comment (TheSwerik) [inline UOP1_Project/Assets/Scripts/UI/FadeManager.cs:11]: Wouldn't just `Fade` be a more appropriate name?
```suggestion
	[Header("Fade Event")]
	[SerializeField] private FadeChannelSO _fadeChannelSO;

	[Header("Fade Canvas")]
```
also change the `[Header("Fade Canvas")]` to `[Header("Fade Image")]` if you apply my other suggestion

Comment (TheSwerik) [reply UOP1_Project/Assets/Scripts/UI/FadeManager.cs:11]: I removed `= default;` because my IDE (Jetbrains Rider) told me this was redundant
A variable will always be set to the default value so `= default` only makes sense if you want to give a function-parameter a default value

Comment (ZombieNub) [reply UOP1_Project/Assets/Scripts/UI/FadeManager.cs:11]: > 
> 
> Wouldn't just `Fade` be a more appropriate name?
> 
> also change the `[Header("Fade Canvas")]` to `[Header("Fade Image")]` if you apply my other suggestion

Will do.

**Review by TheSwerik — APPROVED** [review]
I don't see anything wrong here
Great implementation imo

**Comment by ZombieNub** [comment]
> 
> 
> I don't see anything wrong here
> Great implementation imo

:)

**Comment by ciro-unity** [comment]
Thanks for collaborating on this :)
I like it, I just have a couple of comments which I'm going to fix before merging (no action needed from you, just tell me if I misunderstood something and my changes will break it)

For instance, in the manager you have two functions which are 99% identical (FadeIn and FadeOut), so I unified them.

I'll merge for now. We will need to test it, but also, implement a mechanism so that the scene loading actually doesn't begin until the first fade out has fully finished.
So basically: Disable gameplay controls > Request fade out > Begin scene loading > Scene loading finishes > Begin fade in > Enable gameplay again
